Task 3/Task3.2/Task3.2/DYNAMICARRAY.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+namespace Task3._2
+{
+    class DynamicArray<T> : IEnumerable<T>
+    {
+        internal T[] _array;
+        int _capacity;
+        int _lenght = -1;
+        public T this[int index]
+        {
+            get
+            {
+                if (index > Lenght)
+                    throw new ArgumentOutOfRangeException(nameof(index), "OutOfRangeException");
+                return _array[index];
+            }
+
+            set => _array[index] = value;
+        }
+
+        public int Capacity
+        {
+            get => _capacity;
+            set
+            {
+                _capacity = value;
+                var tempArray = new T[_capacity];
+                for (int i = 0; i < _array.Length; i++)
+                {
+                    tempArray[i] = _array[i];
+                }
+                _array = tempArray;
+            }
+        }
+        public int Lenght 
+        {
+            get => _lenght;
+            private set
+            {
+                _lenght = value;
+                if (Capacity <= _lenght)
+                {
+                    int tempCount = Capacity;
+                    do
+                    {
+                        tempCount *= 2;
+                    } while (tempCount < _lenght);
+                    Capacity = tempCount;
+                }
+            }
+        }
+
+        public DynamicArray()
+        {           
+            _array = new T[8];            
+            Capacity = 8;
+        }
+        public DynamicArray(int Count)
+        {
+            
+            _array = new T[Count];
+            Lenght = Count;
+        }
+        public DynamicArray(IEnumerable<T> item)
+        {
+            _array = new T[0];
+            Capacity = item.Count();
+            AddRange(item);
+        }
+
+
+        public void Add(T item)
+        {
+            Lenght++;
+
+            _array[Lenght] = item;
+
+            //return true;
+        }
+        public bool AddRange(IEnumerable<T> item)
+        {
+            if (Capacity <= _lenght+item.Count())
+            {
+                int tempCount = Capacity;
+                do
+                {
+                    tempCount *= 2;
+                } while (tempCount < _lenght+ item.Count());
+                Capacity = tempCount;
+            }
+            foreach (var i in item)
+            {
+                Add(i);
+            }
+            return true;
```

Comment:
multiple LINQ calls - this causes multiple enumerations. Count is called 3 times - means code had to go through all elements 3 times when it could do that only once.
Call ToArray, to make an array and then use Length at 0 cost. Also there are Array.Copy methods